File: MATLAB_implementation/supporting_functions/newtonnu_vec.m
% ------------------------------------------------------------------------------
%
%                           function newtonnu_vec
%
%  this function finds the mean anomaly given the eccentricity and true anomaly
%  (vectorized version)
%
%  author        : adapted from Vallado for vectorization
%
%  inputs          description                    range / units
%    ecc         - eccentricity                   0.0 to inf (N x 1 vector)
%    nu          - true anomaly                   0.0  to 2pi rad (N x 1 vector)
%
%  outputs       :
%    e0          - eccentric anomaly              0.0  to 2pi rad (N x 1 vector)
%    m           - mean anomaly                   0.0  to 2pi rad (N x 1 vector)
%
%  locals        :
%    small       - tolerance for near-zero values
%    sine        - sine of eccentric anomaly
%    cose        - cosine of eccentric anomaly
%
%  coupling      :
%    none
%
%  references    :
%    vallado       2007, 85-86, alg 5
%
% [e0,m] = newtonnu_vec ( ecc, nu );
% ------------------------------------------------------------------------------

function [e0, m] = newtonnu_vec(ecc, nu)

% -------------------------  implementation   -----------------
small = 0.00000001;

% Initialize output arrays
e0 = zeros(size(ecc));
m = zeros(size(ecc));

% Handle different orbit types
for i = 1:length(ecc)
    if ecc(i) < small
        % Circular orbit
        e0(i) = nu(i);
        m(i) = nu(i);
    elseif ecc(i) < 1.0 - small
        % Elliptical orbit
        sine = (sqrt(1.0 - ecc(i)*ecc(i)) * sin(nu(i))) / (1.0 + ecc(i)*cos(nu(i)));
        cose = (ecc(i) + cos(nu(i))) / (1.0 + ecc(i)*cos(nu(i)));
        e0(i) = atan2(sine, cose);
        m(i) = e0(i) - ecc(i)*sin(e0(i));
    elseif abs(ecc(i) - 1.0) < small
        % Parabolic orbit
        e0(i) = tan(nu(i) * 0.5);
        m(i) = e0(i) + (e0(i)^3)/3.0;
    else
        % Hyperbolic orbit
        sine = (sqrt(ecc(i)*ecc(i) - 1.0) * sin(nu(i))) / (1.0 + ecc(i)*cos(nu(i)));
        e0(i) = asinh(sine);
        m(i) = ecc(i)*sinh(e0(i)) - e0(i);
    end
end

File: python_implementation/supporting_functions/newtonnu_vec.py
"""
Newton-Raphson method for finding mean anomaly from eccentric anomaly (vectorized)

This function converts eccentric anomaly to mean anomaly for elliptical orbits.

Args:
    ecc: eccentricity array
    nu: true anomaly array (radians)

Returns:
    eccentric_anomaly: eccentric anomaly array (radians)
    mean_anomaly: mean anomaly array (radians)
"""

import numpy as np


def newtonnu_vec(ecc, nu):
    """
    Find mean anomaly and eccentric anomaly from true anomaly (vectorized)

    Args:
        ecc: eccentricity array
        nu: true anomaly array (radians)

    Returns:
        tuple: (eccentric_anomaly, mean_anomaly) in radians
    """
    # Ensure inputs are numpy arrays
    ecc = np.atleast_1d(ecc)
    nu = np.atleast_1d(nu)

    # Initialize output arrays
    eccentric_anomaly = np.zeros_like(ecc)
    mean_anomaly = np.zeros_like(ecc)

    # Handle circular orbits (ecc = 0)
    circular = (ecc == 0)
    eccentric_anomaly[circular] = nu[circular]
    mean_anomaly[circular] = nu[circular]

    # Handle elliptical orbits (0 < ecc < 1)
    elliptical = (ecc > 0) & (ecc < 1)
    if np.any(elliptical):
        # Calculate eccentric anomaly from true anomaly
        sqrt_term = np.sqrt((1 - ecc[elliptical]) / (1 + ecc[elliptical]))
        tan_half_nu = np.tan(nu[elliptical] / 2)
        tan_half_e = sqrt_term * tan_half_nu
        eccentric_anomaly[elliptical] = 2 * np.arctan(tan_half_e)

        # Calculate mean anomaly from eccentric anomaly
        e_anom = eccentric_anomaly[elliptical]
        mean_anomaly[elliptical] = e_anom - ecc[elliptical] * np.sin(e_anom)

    # Handle parabolic orbits (ecc = 1)
    parabolic = (ecc == 1)
    if np.any(parabolic):
        # For parabolic orbits, use different formulation
        tan_half_nu = np.tan(nu[parabolic] / 2)
        mean_anomaly[parabolic] = tan_half_nu + (tan_half_nu**3) / 3
        eccentric_anomaly[parabolic] = nu[parabolic]  # No eccentric anomaly for parabolic

    # Handle hyperbolic orbits (ecc > 1)
    hyperbolic = (ecc > 1)
    if np.any(hyperbolic):
        # Calculate hyperbolic eccentric anomaly
        sqrt_term = np.sqrt((ecc[hyperbolic] - 1) / (ecc[hyperbolic] + 1))
        tan_half_nu = np.tan(nu[hyperbolic] / 2)
        tanh_half_h = sqrt_term * tan_half_nu
        eccentric_anomaly[hyperbolic] = 2 * np.arctanh(tanh_half_h)

        # Calculate mean anomaly for hyperbolic orbit
        h_anom = eccentric_anomaly[hyperbolic]
        mean_anomaly[hyperbolic] = ecc[hyperbolic] * np.sinh(h_anom) - h_anom

    return eccentric_anomaly, mean_anomaly
